### PR TITLE
Implements General BulkWriter.

### PR DESCRIPTION
**Pull request by steven2308**
# Description

Describe the additions/improvements this PR contains.

# Actions

Provide overview of actions taken to address the additions/improvements.

# Checklist

- [x] Verified code additions
- [x] Updated NatSpec comments (if applicable)
- [x] Regenerated docs
- [x] Ran prettier
- [x] Added tests to fully cover the changes

---

## PR-Codex overview
This PR adds a new bulk operation error and a sleep function to the RMRK project. It also adds a new general bulk writer contract and modifies the existing one. 

### Detailed summary
- Adds `RMRKCanOnlyDoBulkOperationsOnOwnedTokens()` error to `RMRKErrors.sol`
- Adds `RMRKCanOnlyDoBulkOperationsWithOneTokenAtATime()` error to `RMRKErrors.sol`
- Adds `sleep()` function to `deploy_bulkWriter.ts`
- Adds `RMRKBulkWriter` contract to `RMRKBulkWriter.sol`
- Modifies `RMRKBulkWriterPerCollection` contract to inherit from `RMRKBulkWriter`
- Adds `bulkWriterPerCollection` to `bulkWriterFixture()` in `bulkWriter.ts`
- Adds `bulkWriterPerCollection` to beforeEach() in `Advanced Equip Render Utils` tests
- Adds new tests for `bulkWriterPerCollection` contract

> The following files were skipped due to too many changes: `test/bulkWriter.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by Yuripetusko** [comment]
I was reading about OZ Multicall util (which is totally different thing from MakersDao Multicall3), and was wondering if it can help us here when doing multiple function calls?

OZ  Multicall uses delegatecall and can be used for write functions
MakerDao Multicall uses call and can mostly be used with view functions to aggregate reads

**Comment by steven2308** [comment]
> I was reading about OZ Multicall util (which is totally different thing from MakersDao Multicall3), and was wondering if it can help us here when doing multiple function calls?
> 
> OZ Multicall uses delegatecall and can be used for write functions MakerDao Multicall uses call and can mostly be used with view functions to aggregate reads

As a standalone solution it will not work, since delegate call uses the local memory of the contract with the delegated contracts methods. We would have to add this to the implementation contracts, which will be a problem for size, at least for equippable.

I will see how much it impacts the size.

**Comment by Ayuilos** [comment]
[Lightm](https://github.com/LightmNFT/Lightm-evm) will always offer the best solution of RMRK full functions implementation.

As I discussed with you earlier you can't let contract size become a developer's anxiety, it's a huge mental burden.

**Comment by Ayuilos** [comment]
Where necessary, we should work together to advance the RMRK standard into a composable NFT de facto standard.

Otherwise I am pretty sure you will receive complaints from implementation teams that contract oversize in the future.

**Comment by steven2308** [comment]
> Where necessary, we should work together to advance the RMRK standard into a composable NFT de facto standard.
> 
> Otherwise I am pretty sure you will receive complaints from implementation teams that contract oversize in the future.

It hasn't been our priority but it is coming closer now, I'll make sure to get in touch when we want to add diamond versions to alleviate size issues. Thanks!

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 1]
```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.18;
+
+import "../equippable/IERC6220.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+error RMRKCanOnlyDoBulkOperationsOnOwnedTokens();
+error RMRKCanOnlyDoBulkOperationsWithOneTokenAtATime();
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:9]: I'd extract this to the errors library. I guess there is an argument to be made for keeping the `util` library errors contained there, but on the other hand, it is convenient to have all of the errors in one place 🤔 

TL, DR;  I'd move the errors to the library 😅

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+/**
+ * @title RMRKBulkWriter
+ * @author RMRK team
+ * @notice Smart contract of the RMRK Bulk Writer module.
+ * @dev Extra utility functions for RMRK contracts.
+ */
+contract RMRKBulkWriter {
+    /**
+     * @notice Used to provide a struct for inputing unequip data.
+     * @dev Only used for input and not storage of data
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:20]:
```suggestion
     * @dev Only used for input and not storage of data.
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    struct IntakeUnequip {
+        uint64 assetId;
+        uint64 slotPartId;
+    }
+
+    /**
+     * @notice Reverts if the caller is not the owner of the token
+     * @param collection Address of the collection that this contract is managing.
+     * @param tokenId ID of the token we are managing.
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:32]:
```suggestion
     * @notice Reverts if the caller is not the owner of the token.
     * @param collection Address of the collection that this contract is managing
     * @param tokenId ID of the token we are managing
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    modifier onlyTokenOwner(address collection, uint256 tokenId) {
+        _checkTokenOwner(collection, tokenId);
+        _;
+    }
+
+    /**
+     * @notice Initializes the contract
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:40]:
```suggestion
     * @notice Initializes the contract.
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+     * @dev The `IntakeEquip` stuct contains the following data:
+     *  [
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param collection Address of the collection that this contract is managing.
+     * @param data An `IntakeEquip` struct specifying the equip data.
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:55]:
```suggestion
     * @notice Replaces the current equipped child in the asset and slot combination with the given one.
     * @dev The `IntakeEquip` stuct contains the following data:
     *  [
     *      tokenId,
     *      childIndex,
     *      assetId,
     *      slotPartId,
     *      childAssetId
     *  ]
     * @param collection Address of the collection that this contract is managing
     * @param data An `IntakeEquip` struct specifying the equip data
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            data.slotPartId
+        );
+        IERC6220(collection).equip(data);
+    }
+
+    /**
+     * @notice Performs multiple unequip and equip operations
+     * @dev Unequip operations must run first
+     * @dev tokenId is included as a parameter to be able to do a single check for ownership
+     * @dev Every tokenId in the `IntakeEquip` structs must match the tokenId passed in
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:73]:
```suggestion
     * @notice Performs multiple unequip and/or equip operations.
     * @dev Unequip operations must run first.
     * @dev Unequip operations do not need to be related to the equip operations; this method does not force you to only equip the assets into the slots that were unequipped.
     * @dev `tokenId` is included as a parameter to be able to do a single check for ownership.
     * @dev Every `tokenId` in the `IntakeEquip` structs must match the `tokenId` passed as the argument.
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     *      tokenId,
+     *      childIndex,
+     *      assetId,
+     *      slotPartId,
+     *      childAssetId
+     *  ]
+     * @param collection Address of the collection that this contract is managing.
+     * @param tokenId ID of the token we are managing.
+     * @param unequips[] An array of `IntakeUnequip` structs specifying the slots to unequip.
+     * @param equips[] An array of `IntakeEquip` structs specifying the slots to equip.
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:90]:
```suggestion
     * @param collection Address of the collection that this contract is managing
     * @param tokenId ID of the token we are managing
     * @param unequips[] An array of `IntakeUnequip` structs specifying the slots to unequip
     * @param equips[] An array of `IntakeEquip` structs specifying the slots to equip
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    function bulkEquip(
+        address collection,
+        uint256 tokenId,
+        IntakeUnequip[] memory unequips,
+        IERC6220.IntakeEquip[] memory equips
+    ) public onlyTokenOwner(collection, tokenId) {
+        uint256 length = unequips.length;
+        for (uint256 i = 0; i < length; ) {
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:99]:
```suggestion
        for (uint256 i; i < length; ) {
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                tokenId,
+                unequips[i].assetId,
+                unequips[i].slotPartId
+            );
+            unchecked {
+                ++i;
+            }
+        }
+        length = equips.length;
+        for (uint256 i = 0; i < length; ) {
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:110]:
```suggestion
        for (uint256 i; i < length; ) {
```

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            if (equips[i].tokenId != tokenId) {
+                revert RMRKCanOnlyDoBulkOperationsWithOneTokenAtATime();
+            }
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:113]: I'd add the same check to the unequip actions 🤔

File: contracts/RMRK/utils/RMRKBulkWriter.sol [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                ++i;
+            }
+        }
+    }
+
+    /**
+     * @notice Checks if the caller is the owner of the token
+     * @dev Reverts if the caller is not the owner of the token
+     * @param collection Address of the collection that this contract is managing.
+     * @param tokenId ID of the token we are managing.
```
Comment (ThunderDeliverer) [inline contracts/RMRK/utils/RMRKBulkWriter.sol:125]:
```suggestion
     * @notice Validates that the caller is the owner of the token.
     * @dev Reverts if the caller is not the owner of the token.
     * @param collection Address of the collection that this contract is managing
     * @param tokenId ID of the token we are managing
```

**Review by ThunderDeliverer — CHANGES_REQUESTED** [review]
Nothing major, just a few suggestions.

Reviewing this made me this that BulkWriter might benefit from Smart Contract Tunneling 🤔

File: contracts/RMRK/utils/RMRKBulkWriter.sol (hunk 1 above)
Comment (steven2308) [reply contracts/RMRK/utils/RMRKBulkWriter.sol:9]: Yeah I am more biased towards keeping errors for utils in the utils. But most of them are already on the lib so I will listen to you :grimacing:

File: contracts/RMRK/utils/RMRKBulkWriter.sol (hunk 10 above)
Comment (steven2308) [reply contracts/RMRK/utils/RMRKBulkWriter.sol:114]: Unequips does not have token Id, I take it directly from the param. So nothing to compare there.

Comment (ThunderDeliverer) [reply contracts/RMRK/utils/RMRKBulkWriter.sol:112]: What about the length of `unequips`?

Comment (steven2308) [reply contracts/RMRK/utils/RMRKBulkWriter.sol:112]: What about it?

**Review by ThunderDeliverer — APPROVED** [review]
I only have the comment about the length of `unequips`, but if it can't be done, I'm happy with this PR 😁

**Review by ThunderDeliverer — APPROVED** [review]
I only have the comment about the length of `unequips`, but if it can't be done, I'm happy with this PR 😁

File: contracts/RMRK/utils/RMRKBulkWriter.sol (hunk 10 above)
Comment (ThunderDeliverer) [reply contracts/RMRK/utils/RMRKBulkWriter.sol:112]: Ahhh, brain fart moment 😂

**Pull request merged**